faucet.go

```diff
@@ -232,6 +114,53 @@ func cleanAndExpandPath(path string) string {
 	return filepath.Clean(os.ExpandEnv(path))
 }
 
+// verifyTimeLimit will return an error if the given IP address
+// has already performed an action within the given time limit.
+func verifyTimeLimit(ip string, actionsTimeLimit time.Duration) error {
+	rateLimitMtx.RLock()
+	lastRequestTime, found := requestIPs[ip]
+	rateLimitMtx.RUnlock()
+	if found {
+		nextAllowedRequest := lastRequestTime.Add(actionsTimeLimit)
+		coolDownTime := time.Until(nextAllowedRequest)
+
+		if coolDownTime >= 0 {
+			err := fmt.Errorf("client(%v) may only do an action every "+
+				"%v. Wait another %v.",
+				ip, actionsTimeLimit, coolDownTime)
+			return err
+		}
+	}
+	return nil
+}
+
+// getRealIP returns the clients IP address. If the useRealIP config is set
+// it will use the X-Real-IP or X-Forwarded-For HTTP headers, otherwise it will
+// return the request.RemoteAddr field.
```

Comment:
I suggest making a note/warning here and in config that to use this you need to trust the proxy immediately upstream otherwise the X-Real-IP and X-Forwarded-For headers can be spoofed by a client.